enhancements/automated-resource-request-scaling.md

```diff
@@ -0,0 +1,886 @@
+
+   We would need to change the cluster-version-operator to ignore
+   resource request settings for cluster operators defined in static
+   manifests.
+
+   Allowing an outside actor to change the resource requests turns
+   them into something a user might change themselves, which may not
+   be ideal for our testing and support matrix.
+
+### Options for globally affecting resource limits
+
+We could change some of the underlying components to affect scaling
+without the knowledge of the rest of the cluster components.
+
+1. Have a third-party, such as another operator, change the requests.
+
+   We could investigate using the [vertical pod
+   autoscaler](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler)
+   or write our own operator focused on scaling control plane
+   components.
+
+   If we have 2 controllers trying to manage the same resource (the
+   original operator that owns it and this new scaling operator),
+   there will be contention for the definition. We would need to have
+   all operators ignore the resource requests fields when comparing
+   Deployments of their operands for equality.
+
+   If operators cooperated with the resource request settings, a
+   third-party actor could modify them reliably. We could use an
+   annotation-based API to inform that operator of the base settings
+   for each Deployment, so that during an upgrade if the base needs
+   change they can be scaled appropriately.
+
+2. Have the scheduler and kubelet perform the scaling based on
+   priority class
+
+   Changing kubelet by itself, or adding a plugin, would capture
+   everything as it is launched, but only after a Pod is
+   scheduled. That would mean we would need to leave enough overhead
+   to schedule a large component, even if the request is going to be
+   rewritten to ask for a fraction of that value. So, we have to
+   change the scheduler to ask for less and kubelet to grant less,
+   without changing the Pod definition and triggering the ReplicaSet
+   controller to rewrite the Pod definition.
+
+3. Have kubelet announce more capacity than is available
+
+   We could have kubelet support "over subscription" of a node by
+   reporting that the node has more capacity than it really does. This
+   could be implemented a few ways.
+
+   1. Kubelet could accept an option to report a static amount of extra
+      capacity, for example always reporting 2 more CPUs than are present.
+   2. Kubelet could not subtract the requests from some workloads, or
+      only subtract part of the requests, leaving the capacity higher
+      than it would if the request was included fully in the
+      calculation. *See implementation history above for details about
+      a proof-of-concept implementation of this approach.*
+   3. Kubelet could add capacity to match the requests of some workloads,
+      so that the apparent capacity of a node grows dynamically when
+      system workloads are added.
```

Comment:
Today we discussed this approach, using a regex or list of regexes to identify namespaces to which the behavior would apply. Namespaces give us RBAC, which prevents normal users from abusing the behavior to circumvent quota or DOS the system.